### PR TITLE
Added mongoid 5 support.

### PR DESCRIPTION
**Pull request by dblock**
This is a cleaner version of https://github.com/simi/mongoid_paranoia/pull/37.

* A single Gemfile.
* Tests on more platforms.
* Using mongoid-compatibility to make `update` calls depending on mongoid version.
* Silence debug logger in tests.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,9 +1,21 @@
 source 'https://rubygems.org'
 
-# remove :name arg when migration to '_' gem name is complete
-gemspec name: 'mongoid_paranoia'
+case version = ENV['MONGOID_VERSION'] || '5'
+when /5/
```
Comment (johnnyshields) [inline Gemfile]: Regex should be /^5/. This regex could match MONGOID_VERSION=3.1.5

File: lib/mongoid/paranoia.rb [hunk 2]
```diff
@@ -97,8 +98,12 @@ def persisted?
...
       cascade!
       time = self.deleted_at = Time.now
-      paranoid_collection.find(atomic_selector).
-        update({ "$set" => { paranoid_field => time }})
+      query = paranoid_collection.find(atomic_selector)
+      if Mongoid::Compatibility::Version.mongoid5?
+        query.update_one({ "$set" => { paranoid_field => time }})
+      else
+        query.update({ "$set" => { paranoid_field => time }})
+      end
```
Comment (johnnyshields) [inline lib/mongoid/paranoia.rb]: Might be nice to extract this into a method

File: lib/mongoid/paranoia.rb [hunk 3]
```diff
@@ -145,8 +150,12 @@ def destroyed?
...
     def restore(opts = {})
       run_callbacks(:restore) do
-        paranoid_collection.find(atomic_selector).
-          update({ "$unset" => { paranoid_field => true }})
+        query = paranoid_collection.find(atomic_selector)
+        if Mongoid::Compatibility::Version.mongoid5?
+          query.update_one({ "$unset" => { paranoid_field => true }})
+        else
+          query.update({ "$unset" => { paranoid_field => true }})
+        end
```
Comment (johnnyshields) [inline lib/mongoid/paranoia.rb]: Same here, please extract this into a method

**Comment by simi** [comment]
I think we can delete one gemspec now.

File: Gemfile (hunk 1 above)
Comment (dblock) [inline Gemfile]: Right.

**Comment by simi** [comment]
@dblock since there's only one change needed (`update` -> `update_one`), is there any chance to use `mongoid-compatibility` to handle this? I saw similar changes in more projects.

For example:

```ruby
query.send(Mongoid::Compatibility.update_method, ({ "$unset" => { paranoid_field => true }})
```

Does that make sense?

Another approach can be to ask upstream, if there's a chance to alias that method in new version.

**Comment by dblock** [comment]
I am not 100% sure we want to make mongoid/compatibility a gem that monkey patches behavior, I tried going down that path and I am worried it will create a mess. I'd take PRs there, but in the meantime I think we should make things as explicit as possible.

**Comment by dblock** [comment]
Let me do the removal of mongoid-paranoia.gemspec in a separate PR.

**Comment by simi** [comment]
OK for both comments.

**Comment by dblock** [comment]
Note that `query.send(Mongoid::Compatibility.update_method, ({ "$unset" => { paranoid_field => true }})` is dangerous, because the whole point of mongoid changing `update` to `update_one` and `update_all` is that they wanted to make this a very explicit action, and that the first supported `multi: true` which is no longer possible. So it's much more than a method name.

**Comment by dblock** [comment]
Updated with a private `update` method and a CHANGELOG.

File: mongoid_paranoia.gemspec [hunk 4]
```diff
@@ -17,4 +17,5 @@ Gem::Specification.new do |gem|
   gem.require_paths = ['lib']
 
   gem.add_dependency 'mongoid', '> 3'
```
Comment (simi) [inline mongoid_paranoia.gemspec]: @dblock This should be changed also.

Comment (dblock) [inline mongoid_paranoia.gemspec]: Yep.

**Comment by simi** [comment]
I think this is prepared.

@johnnyshields can you take a look also and merge? I'll bump version, remove old gemspec and release right after this will be merged.

**Comment by dblock** [comment]
I've added a commit that removes the hyphenated .gemspec and clarifies a bit the relationship with that gem (@haihappen said it in a nicer way in his README :)

**Comment by simi** [comment]
Actually this is not true. @haihappen version wasn't working and wasn't 1:1 compatible with 3.0.0 stable extraction.

But if @johnnyshields is ok with this text, let him merge this.

**Comment by dblock** [comment]
I said "the goal is to stay API compatible", not making any claims, just trying to say something nicer than he doesn't want our help ;)

**Comment by simi** [comment]
Let it be. It is better to stay positive. As I said, it's @johnnyshields's turn now. :)

:heart: Thanks for you work @dblock. :heart:

**Comment by johnnyshields** [comment]
@dblock commits are good. Also fine with the new readme wording. For the record, haihappen is a total dick.

**Comment by Aethelflaed** [comment]
Good work @dblock, also my point in not using mongoid compatibility is because I prefer duck typing over version checking. If the API is consistent and has this method, then it should work.

Simple matter of coding-style I suppose. Thanks for the update

**Comment by simi** [comment]
:heart: release is on the way.

**Comment by dblock** [comment]
@Aethelflaed semantically I think mongoid/compatibility is more "correct", because you're saying "i am using the API of version 5 vs. 4", checking whether the method supported is `update` vs. `update_all` can get you in trouble with yet another change with version 6 where suddenly `update` is re-added again or something crazy like that.

**Comment by simi** [comment]
released

**Pull request merged**